File: InterviewPrep/Google/SortingAlgorithms/BubbleSort.py
class Solution:
    def bubbleSort(A):
        n = len(A)
        for i in range(n):
            # The last i elements are already sorted
            for j in range(n-i-1):
                # Swap with the adjacent value if lesser than
                if A[j+1] < A[j]:
                    temp = A[j]
                    A[j] = A[j+1]
                    A[j+1] = temp
        return A

if __name__ == '__main__':
    A = [64, 34, 25, 12, 22, 11, 90]
    print(f"A = {A}")
    S = Solution.bubbleSort(A)
    print(f"A = {A}")
    print(f"S = {S}")

File: Google/SortingAlgorithms/BubbleSort.cpp
#include <iostream>

using namespace std;

void swap(int& a, int& b)
{
    int temp = a;
    a = b;
    b = temp;
}

void bubbleSort(int* A, int n)
{
    for (int i = 0; i < n; i++)
    {
        // The last i elements are already sorted
        for (int j = 0; j < (n - i - 1); j++)
        {
            if (A[j+1] < A[j])
            {
                swap(A[j+1], A[j]);
            }
        }
    }
}

void print(int* A, int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << A[i] << " ";
    }
    cout << endl;
}

int main()
{
    int A[] = {64, 34, 25, 12, 22, 11, 90};
    int n = sizeof(A)/sizeof(A[0]);
    print(A, n);
    bubbleSort(A, n);
    print(A, n);
}
